### PR TITLE
[NavBar] Address QA items

### PR DESCRIPTION
**Pull request by damassi**
As this is a lot of individual QA changes I'll comment inline where appropriate. 

**Zeplin Spec:**
https://app.zeplin.io/project/5acd19ff49a1429169c3128b/screen/5c649937a981283c0b74cc63

**JIRA:**
https://artsyproduct.atlassian.net/browse/PLATFORM-1409#add-comment?atlOrigin=eyJpIjoiOWZlY2ZhOWFlYzM4NDUzMzhjN2ZmZTEzYWZiZDFlMmMiLCJwIjoiaiJ9

#### Feedback check-list: 

- [x] Nav items sit within a flexbox that expands to a max width of 420 (Width will shrink with browser size...390 @ Large, 340 @ Medium)
- [x] Flexbox is fixed & right aligned to login/sign up buttons
- [x] 40px padding to right/left of flexbox
- [x] Search bar expands and fills space
- [x] The amount of nav items will drop at Large breakpoint
- [x] Changed the "More" nav item to text instead of "..." icon
- [x] Removed the "More" subhead within the dropdown
- [x] Visual update to the Activity dropdown zero state
- [x] At Small breakpoint, we switch to the mobile/XS view (Previously did not do this)
- [x] Mobile/XS expanded menu background should extend to bottom of browser (see spec for details)
- [x] [BUG] Mobile/XS menu stays open if you expand your browser. Expected behavior is that it would close/ reset as larger breakpoints are met
- [x] Nav is currently covering the purple banner in staging.
- [x] Some weirdness w/ the white background color not being rounded to match the border, causing a slight overlap in each of the 4 corners. Maybe something that could be addressed with different box-sizing?
- [x] Dropdown placement should align to the right edge of the icon/text it correlates with (see spec for visual) Trying to avoid the dropdown for User/Account running into the right edge of the browser.... I just noticed your comment on Zeplin about this. I'm flexible on this one, we can look at it together.

#### Todo in different PR (part of a separate, larger task) 
- It would be great if the sign-out link would work by default HTML means (eg when JS is broken)

### CODE REVIEW

File: src/Components/NavBar/Menus/MobileNavMenu.tsx [hunk 1]
```diff
@@ -37,6 +37,7 @@ export const MobileNavMenu: React.FC = () => {
       px={2}
       py={1}
       width="100%"
+      height="100vh"
```
Comment (damassi) [inline src/Components/NavBar/Menus/MobileNavMenu.tsx]: Ensure the mobile nav menu fills the screen when expanded

File: src/Components/NavBar/Menus/MoreNavMenu.tsx [hunk 2]
```diff
@@ -19,7 +19,7 @@ export const MoreNavMenu: React.FC = () => {
   }
 
   return (
-    <Menu title="More" onClick={trackClick}>
+    <Menu onClick={trackClick}>
```
Comment (damassi) [inline src/Components/NavBar/Menus/MoreNavMenu.tsx]: Removes the title from the "More" menu

**Comment by artsyit** [comment]
Deploy preview for *artsy-reaction* ready!

Built with commit 75dff7bee5b649ff9eca96ea42a3e5142bf35ed7

https://deploy-preview-2442--artsy-reaction.netlify.com

File: src/Components/NavBar/Menus/NotificationsMenu.tsx [hunk 3]
```diff
@@ -65,20 +66,25 @@ export const NotificationMenuItems: React.FC<
...
-        {notifications.length ? (
-          <>
+        <>
+          {notifications.length === 0 && (
+            <Flex width="100%" flexDirection="column">
+              <Box pt={1} pb={3} width="100%" textAlign="center">
+                <Serif size="3">No new works</Serif>
+              </Box>
+            </Flex>
+          )}
```
Comment (damassi) [inline src/Components/NavBar/Menus/NotificationsMenu.tsx]: Reworks the zero state when no works are available

File: src/Components/NavBar/NavBar.tsx [hunk 4]
```diff
@@ -32,19 +32,28 @@ import * as authentication from "./Utils/authentication"
...
   const [showMobileMenu, toggleMobileNav] = useState(false)
+  const isMobile = useMedia(themeProps.mediaQueries.sm)
   const isLoggedIn = Boolean(user)
 
+  // Close mobile menu if dragging window from small size to desktop
+  useEffect(() => {
+    if (!isMobile) {
+      toggleMobileNav(false)
+    }
+  }, [isMobile])
```
Comment (damassi) [inline src/Components/NavBar/NavBar.tsx]: Adds a useful new hook -- `useMedia` for evaluating when we're at certain breakpoints and addresses issue where when in mobile view with the nav open, and then dragging browser out to desktop, the menu wouldn't close.

File: src/Components/NavBar/NavBar.tsx [hunk 5]
```diff
@@ -64,30 +73,44 @@ export const NavBar: React.FC = track({
         {/*
           Desktop
         */}
-        <NavSection display={["none", "flex"]}>
+        <NavSection display={["none", "none", "flex"]}>
```
Comment (damassi) [inline src/Components/NavBar/NavBar.tsx]: Now, mobile layout is triggered at `sm`, vs `xs`.

File: src/Components/NavBar/NavBar.tsx [hunk 6]
```diff
@@ -102,10 +125,24 @@ export const NavBar: React.FC = track({
...
                 >
-                  <BellIcon top={3} />
+                  {({ hover }) => {
+                    return (
+                      <BellIcon
+                        top={3}
+                        fill={hover ? "purple100" : "black80"}
+                      />
+                    )
+                  }}
```
Comment (damassi) [inline src/Components/NavBar/NavBar.tsx]: We needed a way to pipe UI states into NavItem contents. For example, most nav items are just text so it's easy to change the text color on hover. But occasionally we need to use icons, which require a prop. Using the renderProps pattern we can pipe internal `NavItem` state up into the upper scope to act on changes.

nav

File: src/Components/NavBar/NavItem.tsx [hunk 7]
```diff
@@ -54,7 +57,16 @@ export const NavItem: React.FC<NavItemProps> = ({
...
-          {children}
+          <Box height={25}>
+            {isFunction(children)
+              ? // NavItem children can be called as renderProps so that contents
+                // can operate on UI behaviors (such as changing the color of an
+                // icon on hover).
+                children({
+                  hover,
+                })
+              : children}
```
Comment (damassi) [inline src/Components/NavBar/NavItem.tsx]: This is where the renderProps pattern described in [this comment](https://github.com/artsy/reaction/pull/2442/files#r285838759) is implemented.

File: src/Components/NavBar/NavItem.tsx [hunk 8]
```diff
@@ -64,13 +76,13 @@ export const NavItem: React.FC<NavItemProps> = ({
         </MenuContainer>
       )}
 
-      {showOverlay && <Overlay />}
+      {showOverlay && <Overlay hover={hover} />}
```
Comment (damassi) [inline src/Components/NavBar/NavItem.tsx]: Also pass UI behavior over to `<NavItem Overlay={NotificationBadge} />` components

File: src/Components/NavBar/NavItem.tsx [hunk 9]
```diff
@@ -64,13 +76,13 @@ export const NavItem: React.FC<NavItemProps> = ({
...
     </Box>
   )
 }
 
 const MenuContainer = styled(Box)`
   position: absolute;
-  transform: translate(-80%);
-  padding-top: 18px;
+  transform: translateX(-90%);
+  top: 63px;
```
Comment (damassi) [inline src/Components/NavBar/NavItem.tsx]: Positions the drop-down menus a little more precisely, so they get closer to the right edge of the nav item. Not entirely ideal, pushing things around like but can find a more ideal positioning mechanism in a refactor pass.

File: src/Components/NavBar/NotificationsBadge.tsx [hunk 10]
```diff
@@ -12,9 +11,12 @@ import { NotificationsQueryRenderer } from "./Menus"
...
 
-export const NotificationsBadge: React.FC = () => {
-  const { trackEvent } = useTracking()
-
+export const NotificationsBadge: React.FC<{
+  /**
+   * If hovering over the nav item, `hover` is passed into the badge (Overlay)
+   */
+  hover?: boolean
+}> = ({ hover }) => {
```
Comment (damassi) [inline src/Components/NavBar/NotificationsBadge.tsx]: See overlay comment [here](https://github.com/artsy/reaction/pull/2442/files#r285839025)

File: src/Components/NavBar/NotificationsBadge.tsx [hunk 11]
```diff
@@ -64,35 +66,44 @@ export const NotificationsBadge: React.FC = () => {
...
-          })
-        })
-
         return (
-          <Box onMouseOver={trackOnHover}>
-            <CircularCount notifications={notifications} />
+          <Box>
+            <CircularCount
+              notifications={notifications}
+              rawCount={totalUnread}
```
Comment (damassi) [inline src/Components/NavBar/NotificationsBadge.tsx]: We pass along the raw count in addition to the formatted count. Analytics wants the full integer, not the formatted "99+" label string.

File: src/Components/NavBar/NotificationsBadge.tsx [hunk 12]
```diff
@@ -64,35 +66,44 @@ export const NotificationsBadge: React.FC = () => {
...
   }
 
+  useEffect(() => {
+    if (hover) {
+      trackEvent({
+        subject: AnalyticsSchema.Subject.NotificationBell,
+        new_notification_count: rawCount,
+      })
+    }
+  }, [hover])
```
Comment (damassi) [inline src/Components/NavBar/NotificationsBadge.tsx]: On `hover` change, fire the `useEffect` hook. If `hover` is `true`, track.

File: src/Utils/Hooks/useMedia.ts [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+    const handleChange = event => setMatches(event.matches)
+
+    mediaQueryList.addEventListener("change", handleChange)
+    return () => {
+      mediaQueryList.removeEventListener("change", handleChange)
+    }
+  }, [mediaQueryString])
+
+  return matches
+}
```
Comment (damassi) [inline src/Utils/Hooks/useMedia.ts]: Grabbed this hook from https://github.com/olistic/react-use-media/. Is MIT. Didn't want to have to install a new package for something this simple and something we might change.

File: src/Components/NavBar/NotificationsBadge.tsx [hunk 14]
```diff
@@ -38,61 +40,75 @@ export const NotificationsBadge: React.FC = () => {
...
-            subject: AnalyticsSchema.Subject.NotificationBell,
-            new_notification_count: totalUnread,
-          })
-        })
-
         return (
-          <Box onMouseOver={trackOnHover}>
-            <CircularCount notifications={notifications} />
+          <Box>
+            <CircularCount count={count} rawCount={totalUnread} hover={hover} />
```
Comment (damassi) [inline src/Components/NavBar/NotificationsBadge.tsx]: We pass along the raw count in addition to the formatted `count`. Analytics wants the full integer, not the formatted "99+" label string.

File: src/Components/NavBar/NavBar.tsx [hunk 15]
```diff
@@ -32,19 +32,28 @@ import * as authentication from "./Utils/authentication"
...
 
 export const NavBar: React.FC = track({
   flow: AnalyticsSchema.Flow.Header,
   context_module: AnalyticsSchema.ContextModule.Header,
 })(_props => {
-  const { mediator, user } = useContext(SystemContext)
   const { trackEvent } = useTracking()
+  const { mediator, user } = useContext(SystemContext)
   const [showMobileMenu, toggleMobileNav] = useState(false)
+  const isMobile = useMedia(themeProps.mediaQueries.sm)
```
Comment (mzikherman) [inline src/Components/NavBar/NavBar.tsx]: nice!

File: src/Components/NavBar/NotificationsBadge.tsx [hunk 16]
```diff
@@ -38,61 +40,75 @@ export const NotificationsBadge: React.FC = () => {
...
-}) => {
+const CircularCount: React.FC<{
+  /**
+   * Formatted count for display
+   */
+  count?: string
+  /**
+   * Raw unread count, used for analytics.
+   */
+  rawCount?: boolean
```
Comment (mzikherman) [inline src/Components/NavBar/NotificationsBadge.tsx]: Is this `number`?

Comment (damassi) [reply src/Components/NavBar/NotificationsBadge.tsx]: Weird! how did this slip through

Comment (damassi) [reply src/Components/NavBar/NotificationsBadge.tsx]: Ahh, I know how -- the fallback, 0, is falsy:

Screen Shot 2019-05-21 at 11 58 18 AM

Fixed by adding a type

**Comment by artsyit** [comment]
:rocket: PR was released in v16.12.5 :rocket:

**Pull request merged**